File: org.eclipse.examples.toast.rap.widgets.drivingsimulator/src/org/eclipse/examples/toast/rap/drivingsimulator/js/DrivingSimulator.js
/*******************************************************************************
 * Copyright (c) 2009 EclipseSource Inc. and others. All rights reserved.
 * This program and the accompanying materials are made available under the 
 * terms of the Eclipse Public License v1.0 and Eclipse Distribution License v1.0 
 * which accompanies this distribution. The Eclipse Public License is available at 
 * http://www.eclipse.org/legal/epl-v10.html and the Eclipse Distribution License 
 * is available at http://www.eclipse.org/org/documents/edl-v10.php.
 *
 * Contributors: 
 *     EclipseSource Inc. - initial API and implementation
 *******************************************************************************/
qx.Class.define( "org.eclipse.examples.toast.rap.drivingsimulator.DrivingSimulator", {
  extend: qx.ui.layout.CanvasLayout,
  
  construct: function() {
    this.base( arguments );
    DS_widget = this;
    this.addEventListener( "appear", this._loadEarth, this );
  
    this.timer.addEventListener("interval", function(e) {
      if( !org_eclipse_rap_rwt_EventUtil_suspend ) {
          if( DS_ge != null ) {
            var wm = org.eclipse.swt.WidgetManager.getInstance();
            var sim = wm.findIdByWidget( DS_widget );
            var req = org.eclipse.swt.Request.getInstance();
            req.addParameter( sim + ".wantFly", true );
            req.send();
          }
      }
        
     }, this );      
     this.timer.start(); 
     
     this.addEventListener( "blur", function(e){
       this.timer.stop();
       DS_ge = null;
     }, this );
    
  },
  
    
  members : {
    timer : new qx.client.Timer( 5000 ),
    wantLoad : false,
    wantStart : false,
    wantFly : false,
    latitude : 0,
    longitude : 0,
    heading : 0,
    zoom : 0,
    oLat : 0,
    oLong : 0,
    dLat : 0,
    dLong : 0,

    _loadEarth : function() {
      if( DS_ge == null ) {
        google.earth.createInstance( this.getElement(), 
                                     this._initCB, 
                                     this._failureCB );
        mapParent = document.createElement('div');
  
        DS_map = new GMap2( mapParent );           
      }                            
    },
    
    _doActivate : function() {
      var shell = null;
      var parent = this.getParent();
      while( shell == null && parent != null ) {
        if( parent.classname == "org.eclipse.swt.widgets.Shell" ) {
          shell = parent;
        }
        parent = parent.getParent();
      }
      if( shell != null ) {
        shell.setActiveChild( this );
      }
     },
      
    _initCB : function( object ) {
      DS_ge = object;
      DS_ge.getWindow().setVisibility( true );
      DS_ge.getLayerRoot().enableLayerById( DS_ge.LAYER_BUILDINGS, true );
      DS_ge.getLayerRoot().enableLayerById( DS_ge.LAYER_BORDERS, true );
      DS_geHelpers = new GEHelpers( DS_ge );
      
      if( DS_widget.wantLoad ) {
        DS_widget.wantLoad = false;
        DS_widget.loadDirections( DS_widget.oLat, 
                                  DS_widget.oLong, 
                                  DS_widget.dLat, 
                                  DS_widget.dLong );  
      } 
      if( DS_widget.wantFly ) {
        DS_widget.flyTo( DS_widget.latitude,
                         DS_widget.longitude,
                         DS_widget.heading,
                         DS_widget.zoom );
        DS_widget.wantFly = false;                         
      }
    },
    
    _failureCB : function() {
      alert('loading earth failed, maybe its destroyed?');
    },
    
    startDriving : function() {
      if( DS_ge != null ) {
        DS_controlSimulator( 'start' );
      } else {
        DS_widget.wantStart = true;
      }
    },
    
    stopDriving : function() {
      if( DS_ge != null ) {
        DS_controlSimulator( 'pause' );
      }
    },
    
    increaseSpeed : function() {
      if( DS_ge != null ) {
        DS_controlSimulator( 'faster' );
      }
    },
    
    decreaseSpeed : function() {
      if( DS_ge != null ) {
        DS_controlSimulator( 'slower' );
      }
    },
    
    reset : function() {
      if( DS_ge != null ) {
        DS_controlSimulator( 'reset' );
      }
    },
    
    flyTo : function( latitude, longitude, heading, zoom ) {
      DS_widget.latitude = latitude;
      DS_widget.longitude = longitude;
      DS_widget.heading = heading;
      DS_widget.zoom = zoom;
      if( DS_ge != null ) {
        lats = latitude / 1000000.0;
        longs = longitude / 1000000.0;
        DS_flyToLatLng2( new google.maps.LatLng( lats, longs ), heading, zoom); 
      } else {
        DS_widget.wantFly = true;
      }
    },
    
    loadDirections : function( oLat, oLong, dLat, dLong ) {
      DS_widget.oLat = oLat;
      DS_widget.oLong = oLong;
      DS_widget.dLat = dLat;
      DS_widget.dLong = dLong;
      if( DS_ge != null ) {
        oLat2 = DS_widget.oLat / 1000000.0;
        oLong2 = DS_widget.oLong / 1000000.0;
        dLat2 = DS_widget.dLat / 1000000.0;
        dLong2 = DS_widget.dLong / 1000000.0;
        DS_goDirections2( oLat2, oLong2, dLat2, dLong2);
      } else {
        DS_widget.wantLoad = true;
      }
    },
    
    _doResize : function() {
      //qx.ui.core.Widget.flushGlobalQueues();
    }      
    
  }

} );   

File: org.eclipse.examples.toast.rap.widgets.drivingsimulator/src/org/eclipse/examples/toast/rap/drivingsimulator/DrivingSimulator.java
/*******************************************************************************
 * Copyright (c) 2009 EclipseSource Inc. and others. All rights reserved.
 * This program and the accompanying materials are made available under the 
 * terms of the Eclipse Public License v1.0 and Eclipse Distribution License v1.0 
 * which accompanies this distribution. The Eclipse Public License is available at 
 * http://www.eclipse.org/legal/epl-v10.html and the Eclipse Distribution License 
 * is available at http://www.eclipse.org/org/documents/edl-v10.php.
 *
 * Contributors: 
 *     EclipseSource Inc. - initial API and implementation
 *******************************************************************************/
package org.eclipse.examples.toast.rap.drivingsimulator;

import org.eclipse.swt.widgets.Composite;

/**
 * This class represents the Driving Simulator widget. It creates a google
 * earth representation on the client site. You can fly to positions or 
 * calculate a route. After calculating a rout you can send a Smart on the 
 * trip. Works only with if the Google Earth Browser Plugin is installed. If
 * not see: http://code.google.com/intl/de-DE/apis/earth/documentation/#install 
 */
public class DrivingSimulator extends Composite {
  
  private DrivingSimulatorAdapter adapter;
  private int[] coordinates;
  private boolean decreaseSpeed;
  private int[] flyData;
  private boolean flyTo;
  private boolean increaseSpeed;
  private boolean readyToStart;
  private boolean reset;
  private boolean start;
  private boolean stop;
  
  public DrivingSimulator( Composite parent, int style ) {
    super( parent, style ); 
    adapter = new DrivingSimulatorAdapter( this );
    readyToStart = false;
    start = false;
    stop = false;
    increaseSpeed = false;
    decreaseSpeed = false;
    reset = false;
    flyTo = false;
  }
  
  /**
   * Decrease the driving speed of the Smart
   */
  public void decreaseSpeed() {
    decreaseSpeed = true;
  }

  /**
   * Move the users view to the given location.
   */
  public void flyTo(
    final int latitude, 
    final int longitude, 
    final int heading, 
    final int zoom) 
  {
    flyData = new int[] { latitude, longitude, heading, zoom };
    flyTo = true;
  }
  
  public Object getAdapter( Class adapter ) {
    Object result = null;
    if( adapter == DrivingSimulatorAdapter.class ) {
      result = this.adapter;
    } else {
      result = super.getAdapter( adapter );
    }
    return result;
  }
  
  /**
   * Returns the starting and endpoint lat/long as an array.
   * the 1st and 2nd entries are the origin lat/long. The 3rd and 4th entries 
   * are the target lat/long.
   * 
   * @return The route points as an array.
   */
  public int[] getDirections() {
    return coordinates;
  }
  
  int[] getFlyData() {
    return flyData;
  }
  
  boolean getReadyToStart() {
    return readyToStart;
  }
  
  /**
   * Increase the driving speed of the Smart
   */
  public void increaseSpeed() {
    increaseSpeed = true;
  }
  
  boolean isDecreaseSpeed() {
    return decreaseSpeed && readyToStart;
  }
  
  /**
   * Check if the Smart is driving.
   * @return <code>true</code> when the Smart drives.
   */
  public boolean isDriving() {
    return start && readyToStart;
  }
  
  boolean isFlyTo() {
    return flyTo;
  }
  
  boolean isIncreaseSpeed() {
    return increaseSpeed && readyToStart;
  }

  boolean isReset() {
    return reset && readyToStart;
  }
  
  /**
   * Calculates the route from the origin lat/long the the destination
   * lat/long.
   * @param originLatitude The starting latitude.
   * @param originLongitude The starting longitude.
   * @param destLatitude The target latitude.
   * @param destLongitude The target longitude.
   */
  public void loadDirections( 
    final int originLatitude, 
    final int originLongitude, 
    final int destLatitude, 
    final int destLongitude) 
  {
    coordinates = new int[] { originLatitude, originLongitude, 
                              destLatitude, destLongitude };
    readyToStart = true;
  }
  
  /**
   * Reset the the Smart to it's starting point.
   */
  public void reset() {
    reset = true;
  }
  
  void resetDecreaseSpeed() {
    decreaseSpeed = false;
  }
  
  void resetDone() {
    reset = false;
  }
  
  void resetFlyTo() {
    flyTo = false;
  }
  
  void resetIncreaseSpeed() {
    increaseSpeed = false;
  }
  
  void resetStart() {
    start = false;
  }
  
  void resetStop() {
    stop = false;
  }
  
  /**
   * Let the Smart drive.
   */
  public void start() {
    if( readyToStart && !start ) {
      start = true;
    }
  }
  
  /**
   * Stops the Smart if it's driving.
   */
  public void stop() {
    if( readyToStart && !stop ) {
      stop = true;
    }
  }
  
  boolean wantToStart() {
    return readyToStart && start;
  }
  
  boolean wantToStop() {
    return readyToStart && stop;
  }

}
